File: android/src/main/kotlin/jyahann/dgis_map_kit/ClusterRenderer.kt
package jyahann.dgis_map_kit

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.MethodChannel
import ru.dgis.sdk.map.SimpleClusterObject
import ru.dgis.sdk.map.SimpleClusterOptions
import ru.dgis.sdk.map.SimpleClusterRenderer
import java.util.concurrent.CompletableFuture

class ClusterRenderer(methodChannel: MethodChannel, sdkContext: ru.dgis.sdk.Context, layerId: String?) :
        SimpleClusterRenderer {
    private var methodChannel: MethodChannel
    private var sdkContext: ru.dgis.sdk.Context
    private var layerId: String?

    init {
        this.methodChannel = methodChannel
        this.sdkContext = sdkContext
        this.layerId = layerId;
    }

    override fun renderCluster(cluster: SimpleClusterObject): SimpleClusterOptions {
        val markers: MutableList<Any?> = ArrayList()
        for (dgisObject in cluster.objects) {
            var mapUserData = dgisObject.userData as MapObjectUserData

            markers.add(mapUserData.userData)
        }

        var result = ClusterRenderResult(sdkContext, markers, layerId)

        Handler(Looper.getMainLooper()).post {
            methodChannel.invokeMethod(
                "cluster#render",
                mapOf<String, Any?>(
                    "layerId" to layerId,
                    "data" to markers
                ),
                result
            )
        }

        return result.completableClusterOptions.get()
    }
}

class ClusterRenderResult(sdkContext: ru.dgis.sdk.Context, markers: List<Any?>, layerId: String?) : MethodChannel.Result {
    var completableClusterOptions = CompletableFuture<SimpleClusterOptions>()

    private var sdkContext: ru.dgis.sdk.Context
    private var markers: List<Any?>
    private var layerId: String?

    init {
        this.sdkContext = sdkContext
        this.markers = markers
        this.layerId = layerId
    }

    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?): Unit {}

    override fun success(result: Any?): Unit {
        completableClusterOptions.complete(
                MarkersUtils.getClusterOptionsFromDart(
                        result as Map<String, Any>,
                        markers,
                        sdkContext,
                        layerId
                )
        )
    }

    override fun notImplemented(): Unit {}
}

File: ios/Classes/ClusterRenderer.swift
import DGis;
import Flutter;

final class ClusterRenderer: SimpleClusterRenderer {
    private var methodChannel: FlutterMethodChannel;
    private var registrar: FlutterPluginRegistrar;
    private var sdk: DGis.Container;
    private var layerId: String?;
    
    init(methodChannel: FlutterMethodChannel, registrar: FlutterPluginRegistrar, sdk: DGis.Container, layerId: String? = nil) {
        self.methodChannel = methodChannel
        self.registrar = registrar
        self.sdk = sdk
        self.layerId = layerId
    }
    
    func renderCluster(cluster: SimpleClusterObject) -> SimpleClusterOptions {
        var clusterOptions: SimpleClusterOptions?;
        let semaphore = DispatchSemaphore(value: 0);
        var markers = Array<Any>();
        
        for object in cluster.objects {
            let userData = object.userData as! MapObjectUserData;
            
            markers.append(userData.userData!);
        }
        
        let args: Dictionary<String, Any?> = [
            "layerId": self.layerId,
            "data": markers
        ];
        
        DispatchQueue.main.async {
            self.methodChannel.invokeMethod(
                "cluster#render",
                arguments: args,
                result: { result in
                    clusterOptions = MarkersUtils.getClusterOptionsFromDart(
                        mapClusterer: result as! Dictionary<String, Any?>,
                        sdk: self.sdk,
                        registrar: self.registrar,
                        markers: markers,
                        layerId: self.layerId
                    )
                    semaphore.signal();
                }
            )
        }
        
        semaphore.wait();
            
        return clusterOptions!;
    }
}

